### ISSUE TITLE
S3 Backup using Minio - Unable to get Secret

### ISSUE THREAD

**Issue by damienbutt** [longhorn/longhorn#2901]
Hi all,

I am trying to get Longhorn backups setup but having a little trouble. I have a Minio server running on TrueNAS at http://192.168.10.83:9000. I have an access ID and an access secret key, and using those am I able to login to the web interface. I have created a bucket called "longhorn". The URL I actually use to access the Minio server is https://minio.domain.com as I am running it behind HAProxy on pfSense which does my SSL termination using Let's Encrypt.

I have added the secret to the Longhorn-System namespace with the base64 encoded AWS_ACCESS_ID, AWS_ACCESS_SECRET_KEY, and AWS_ENPOINTS key/value pairs. 

The URL I base64 encoded was https://minio.domain.com. 

In the Longhorn settings, I set the backup target as s3://longhorn@minio.domain.com/ and then add the name of the secret in the secret field beneath. 

Not sure what I am doing wrong, but it's not working. It keeps saying "error listing backups: cannot backup: unable to get settings backup-target-credential-secret" when I try to select the backups menu. 

Any help would be greatly appreciated. 

**Environment: K3S and Rancher
 - Longhorn version: 1.1.1
 - Kubernetes version: 1.21.3+k3s1
 - Node config: 2x master/control-plane nodes running behind 1x NGINX load balancer, 4x worker nodes, 1x MYSQL external database.
   - OS type and version: Ubuntu Server 20.04 LTS
   - CPU per node: 8
   - Memory per node: 16GB
   - Disk type: SCSI 
   - Network bandwidth and latency between the nodes: 10G-Base-T
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Proxmox/KVM

Thanks,

D

**Comment by damienbutt** [comment]
I should probably also mention that this is my first Kubernetes cluster so I'm a bit of a noob, which is why I have set all this up to have a play around with the tech. I am a fairly technical person though. It's nothing mission-critical, just a home lab environment. Learning it from reading docs, and trial and error.

**Comment by c3y1huang** [comment]
@damienbutt thank you for the detailed issue description.
Do you see a value with `kubectl -n longhorn-system get setting/backup-target-credential-secret`?

**Comment by damienbutt** [comment]
Hi,

NAME                                                    VALUE       AGE
backup-target-credential-secret                          20h

No, there doesn't seem to be a value.

**Comment by damienbutt** [comment]
If I run `kubectl get secrets -n longhorn-system`, I see my secret there. "minio-secret". That iswhat I am putting in the backup-target-credential-secret in Longhorn.

<img width="970" alt="Screenshot 2021-08-19 at 12 31 57" src="https://user-images.githubusercontent.com/22627489/130061487-e1b1461f-9b26-4c90-bc84-9aa54e2017dd.png">

**Comment by c3y1huang** [comment]
Not sure why the secret did not save to the setting. Can you try to manually update the `minio-secret` to the setting value?
```
kubectl -n longhorn-system edit setting/backup-target-credential-secret
```

**Comment by damienbutt** [comment]
Hi,

I have actually got it working (I think), in a different way from what I wanted. There are no errors now when I access the Backups menu in Longhorn now.

After a fair bit of trial and error, what I have had to do though is use `s3://longhorn@192.168.10.83:9000/` as my backup target. And for my minio-secret, I entered the ACCESS_ID and SECRET_KEY, and ENDPOINT without base64 encoding them first. Not sure where I got that from? I may have misread the docs. I also had to specify the ENDPOINT in the secret as `http://192.168.10.83:9000/`. 

During the trial and testing, the minio-secret must have saved properly as I didn't have to run the command you provided above and if I now run `kubectl -n longhorn-system get setting/backup-target-credential-secret`, I get:
<img width="586" alt="Screenshot 2021-08-20 at 13 36 10" src="https://user-images.githubusercontent.com/22627489/130234014-421830c0-1963-4959-8c4f-3165d97d7820.png">

But when trying to use `s3://longhorn@minio.domain.com/` as the target and `https://minio.domain.com/` as my ENDPOINT in the secret, it was giving me a very long error which seemed to be the same error repeated multiple times mentioning missing delimiter, missing prefix etc, but at the end an error above not be able to resolve the URL from the 10. subnet that Kubernetes is using. My conclusion eventually was that it was just a DNS problem. Changing to the HTTP URL using the direct IP address and port number worked straight away. 

I looked around the settings but couldn't find a way to tell Kubernetes or Rancher about my internal DNS servers which would resolve the HTTPS URL. Is there a way to do this? I have 2x internal DNS servers running on 192.168.10.251 and 192.168.10.252. Ideally, I would like to run it over HTTPS if I can. 

The HTTPS URL does work as I am able to use it successfully with the Minio MC client to manage the server from my laptop. But my laptop is obviously able to resolve DNS from my internal DNS servers.

**Comment by c3y1huang** [comment]
Longhorn provide an HTTPS sample for [Using a self-signed SSL certificate for S3 communication
](https://longhorn.io/docs/1.1.1/snapshots-and-backups/backup-and-restore/set-backup-target/#using-a-self-signed-ssl-certificate-for-s3-communication).
However, AFAIK Longhorn does not handle the DNS part, you could ask the rancher folks about this.

**Comment by jenting** [comment]
I think the problem has been resolved since the DNS part isn't controller by Longhorn.
I'll close the issue now, feel free to reopen it if you have further questions.

**Comment by sethbrasile** [comment]
@damienbutt I just experienced the same thing. I think I got the idea that you needed to base64 encode secrets for longhorn from a couple of different places, one being here: https://www.civo.com/learn/backup-longhorn-volumes-to-a-minio-s3-bucket

I fussed around with this for hours and your comment was the only thing I found that made clear what was happening and got me up and running so thank you.

Some keyword stuffing hoping to assist future internet searchers: local non-aws url minio longhorn secret base64 encoding encoded no such host

What made it extra hard was that when setting the endpoint URL to a plain text value, longhorn wouldn't load the new secret, it would throw the previous error, which felt like it confirmed the need to base64 encode the values. It wasn't until I removed base64 encoding from all of the values at once that it would try with a plain text URL.

For the future reader, what ended up working for me was a bucket named "longhorn" and a Service Account (for the access key and access key ID) in minio, plain text values in the secret, `AWS_ENDPOINTS` value of `http://<local ip>:9000` and a "backup target" value of `s3://longhorn@local/` (region doesn't appear to matter since this is a single node setup I guess? it works with a few different values, seems to still work as `s3://longhorn@butts/` but have not tested running a backup yet)
